tracee-ebpf/main.go

```diff
@@ -958,42 +952,50 @@ func prepareEventsToTrace(eventFilter *tracee.StringFilter, setFilter *tracee.St
 	return res, nil
 }
 
-var necessaryNewCapabilities = []capability.Cap{capability.CAP_IPC_LOCK, capability.CAP_SYS_PTRACE, capability.CAP_SYS_RESOURCE, capability.CAP_BPF, capability.CAP_PERFMON}
-var necessaryOldCapabilities = []capability.Cap{capability.CAP_IPC_LOCK, capability.CAP_SYS_PTRACE, capability.CAP_SYS_RESOURCE, capability.CAP_SYS_ADMIN}
+const bpfCapabilitiesMinKernelVersion = "5.8"
 
-func dropUnnecessaryCapabilities(caps capability.Capabilities) error {
+func ensureCapabilities() error {
+	selfCap, err := getSelfCapabilities()
+	if err != nil {
+		return err
+	}
 	OSInfo, err := helpers.GetOSInfo()
 	if err != nil {
 		return err
 	}
-	var necessaryCapabilities []capability.Cap
-	if OSInfo.CompareOSBaseKernelRelease("5.8") <= 0 {
-		necessaryCapabilities = necessaryNewCapabilities
+	requiredCapabilities := []capability.Cap{capability.CAP_IPC_LOCK, capability.CAP_SYS_PTRACE, capability.CAP_SYS_RESOURCE}
+	if OSInfo.CompareOSBaseKernelRelease(bpfCapabilitiesMinKernelVersion) <= 0 {
+		requiredCapabilities = append(requiredCapabilities, []capability.Cap{capability.CAP_BPF, capability.CAP_PERFMON}...)
 	} else {
-		necessaryCapabilities = necessaryOldCapabilities
+		requiredCapabilities = append(requiredCapabilities, []capability.Cap{capability.CAP_SYS_ADMIN}...)
 	}
-	caps.Clear(capability.CAPS)
-	for _, ncap := range necessaryCapabilities {
-		caps.Set(capability.CAPS, ncap)
+
+	if err = checkRequiredCapabilities(selfCap, requiredCapabilities); err != nil {
+		return err
 	}
-	err = caps.Apply(capability.CAPS)
-	if err != nil {
-		return fmt.Errorf("couldn't drop capabilities: %v", err)
+	if err = dropUnnecessaryCapabilities(selfCap, requiredCapabilities); err != nil {
+		return err
 	}
 	return nil
 }
 
-func checkRequiredCapabilities(caps capability.Capabilities) error {
-	if !caps.Get(capability.EFFECTIVE, capability.CAP_SYS_ADMIN) {
-		return fmt.Errorf("insufficient privileges to run: missing CAP_SYS_ADMIN")
+func dropUnnecessaryCapabilities(selfCaps capability.Capabilities, necessaryCapabilities []capability.Cap) error {
```

Comment:
can't you call neededCapabilities to callees and callers ? You're calling requireCapabilities in some places, necessaryCapabilities in some others. Feels confusing to me.

Reply:
You are right, it is confusing.
I think I will use required, because I thinks it describes the best that this is a setup for the code requirements.